Telegram bot support

kpaveliev opened the pull request:
### Description

Telegram bot support added. Now users can connect their telegram account to the app and:

- get list of their goals
- add new goals with title and category

### Additions to the app

- external packages: marshmallow_dataclass, marshmallow
- bot app with the all logic
- bot app: TgUser model to keep telegram users ids, chat ids, match app and telegram users
- bot app: TgUserUpdateView (/bot/verify) to verify users using front
- bot app: telegram API dataclasses to keep telegram responses
- bot app: TgClient to communicate with telegram API
- bot app: runbot manage.py command (key logic is here)
- goals app: Goal model updated to allow null values in due_date, description fields
- deploy: front version update
- deploy: additional container to communicate with the bot
- .env, github actions, github secrets: TG_TOKEN from the bot

kpaveliev committed:
bot: app created

kpaveliev committed:
bot: app added to settings and urls

kpaveliev committed:
bot: TgUser model created

kpaveliev committed:
bot: TgUser model added to admin

kpaveliev committed:
bot: telegram API dataclasses draft

kpaveliev committed:
bot: requirements update - marshmallow, marshmallow-dataclass

kpaveliev committed:
bot: telegram API dataclassed descriptions added

kpaveliev committed:
bot: TgClient get_updates draft

kpaveliev committed:
bot: TgClient get_updates method and changes to dataclasses

kpaveliev committed:
bot: TgClient send_message added + dataclass changes

kpaveliev committed:
bot: marshmallow_dataclass @dataclass decorator to avoid generating schemas explicitly

kpaveliev committed:
bot: runbot command added, .env and settings updated to add TG_TOKEN

kpaveliev committed:
bot-runbot: bot sends same message as reply

kpaveliev committed:
bot-runbot: different message in case user already wrote or new

kpaveliev committed:
bot-runbot: verification_code added

kpaveliev committed:
bot-runbot: verification code sent to new users

kpaveliev committed:
bot-views: bot/verify update view added

kpaveliev committed:
bot-views: working update view for verification

kpaveliev committed:
bot-views: working update view for verification

kpaveliev committed:
bot-runbot: bot returns list of goals if user is authorized

kpaveliev committed:
bot-runbot: message for unknown command

kpaveliev committed:
bot-deploy: bot container added

kpaveliev committed:
bot-deploy: front image updated

kpaveliev committed:
bot-deploy: TK_TOKEN secret added to actions

kpaveliev committed:
bot-deploy: deploy test

kpaveliev committed:
bot-deploy: bot troubleshooting #1

kpaveliev committed:
bot-deploy: bot troubleshooting #2

kpaveliev committed:
bot-deploy: bot troubleshooting #3 - validation error arises periodically

kpaveliev committed:
bot-deploy: bot troubleshooting #3 - validation error arises periodically

kpaveliev committed:
bot-runbot: refactor

kpaveliev committed:
bot-runbot: all commands logic draft - refactor needed

kpaveliev committed:
bot: Goal model updated - allow null in due_date, description fields

kpaveliev committed:
bot-runbot: refactor

kpaveliev committed:
bot-runbot: refactors - logic distributed between methods

kpaveliev committed:
bot: README update

kpaveliev committed:
bot-runbot: refactor - only categories user own or has permission to write + bugs

kpaveliev committed:
bot: deploy test

kpaveliev committed:
bot-runbot: cancel message changed

kpaveliev committed:
README: bot username added

painassasin commented on todolist/bot/management/commands/runbot.py:
```diff
@@ -0,0 +1,159 @@
+            board__participants__user=self.user.user,
+            board__participants__role__in=[BoardParticipant.Role.owner, BoardParticipant.Role.writer],
+            is_deleted=False
+        ).all()
```
Думаю тут лучше использовать `only`, чтобы не тащить все поля модели, ведь тебе нужны только `id` и `title`

kpaveliev replied:
Учтено

painassasin commented on todolist/bot/management/commands/runbot.py:
```diff
@@ -0,0 +1,159 @@
+
+    def _send_reply(self, reply: str | list) -> None:
+        """Send reply"""
+        if type(reply) == list:
```
Типы лучше сравнивать через `isinstance`
```python
if isinstance(reply, list):
    ...
```

kpaveliev replied:
Учтено

kpaveliev committed:
bot-runbot: bug fix - same reply every minute for last user message

kpaveliev committed:
bot-runbot: bug fix - same reply every minute for last user message

kpaveliev committed:
bot-runbot: bug fix - same reply every minute for last user message

kpaveliev committed:
bot-runbot: bug fix - same reply every minute for last user message

kpaveliev committed:
PR review comments accepted:
- use only() when obtaining goals, categories to get only id, title fields
- use isinstance to check types
